--- /mnt/disk4/tmp/OpenSTLinux/Distribution-kit/build-openstlinuxweston-stm32mp1/conf/bblayers.conf	2018-10-18 15:37:23.792569319 +0200
+++ bblayers.conf	2018-11-19 14:38:51.357263001 +0100
@@ -61,6 +61,9 @@
     if os.path.isfile('${OEROOT}/meta-security/conf/layer.conf') else ''} \
     "
 
+# Av96
+ADDONSLAYERS += "${@'${OEROOT}/meta-st/meta-av96' if os.path.isfile('${OEROOT}/meta-st/meta-av96/conf/layer.conf') else ''}"
+
 # specific to framework
 FRAMEWORKLAYERS += "${@'${OEROOT}/meta-st/meta-st-openstlinux' if os.path.isfile('${OEROOT}/meta-st/meta-st-openstlinux/conf/layer.conf') else ''}"
 
