diff -Naurw ./fdts/stm32mp157a-av96.dts ../git/fdts/stm32mp157a-av96.dts
--- ./fdts/stm32mp157a-av96.dts	1970-01-01 01:00:00.000000000 +0100
+++ ../git/fdts/stm32mp157a-av96.dts	2018-11-28 10:23:06.324567396 +0100
@@ -0,0 +1,25 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause) */
+/*
+ * Copyright (C) STMicroelectronics 2017 - All Rights Reserved
+ * Author: Ludovic Barre <ludovic.barre@st.com> for STMicroelectronics.
+ */
+
+/dts-v1/;
+#include "stm32mp157c-ed1.dts"
+
+/ {
+	model = "STMicroelectronics STM32MP157C-EV1 pmic eval daughter on eval mother";
+	compatible = "st,stm32mp157c-ev1", "st,stm32mp157c-ed1", "st,stm32mp157";
+
+	chosen {
+		bootargs = "earlyprintk console=ttyS3,115200 root=/dev/ram";
+		stdout-path = "serial3:115200n8";
+	};
+};
+
+&usart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&usart3_pins_a>;
+	resets = <&rcc USART3_R>;
+	status = "disabled";
+};
diff -Naurw ./fdts/stm32mp157c-ed1.dts ../git/fdts/stm32mp157c-ed1.dts
--- ./fdts/stm32mp157c-ed1.dts	2018-11-22 16:32:03.000000000 +0100
+++ ../git/fdts/stm32mp157c-ed1.dts	2018-11-28 10:23:06.324567396 +0100
@@ -106,6 +106,7 @@
 	status = "okay";
 };
 
+#if 0
 &sdmmc2 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&sdmmc2_b4_pins_a &sdmmc2_d47_pins_a>;
@@ -117,6 +118,7 @@
 	bus-width = <8>;
 	status = "okay";
 };
+#endif
 
 &uart4 {
 	pinctrl-names = "default";
diff -Naurw ./fdts/stm32mp157c-ev1.dts ../git/fdts/stm32mp157c-ev1.dts
--- ./fdts/stm32mp157c-ev1.dts	2018-11-22 16:32:45.000000000 +0100
+++ ../git/fdts/stm32mp157c-ev1.dts	2018-11-28 10:23:06.324567396 +0100
@@ -17,9 +17,11 @@
 	};
 };
 
+#if 0
 &usart3 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&usart3_pins_a>;
 	resets = <&rcc USART3_R>;
 	status = "disabled";
 };
+#endif
diff -Naurw ./fdts/stm32mp157-pinctrl.dtsi ../git/fdts/stm32mp157-pinctrl.dtsi
--- ./fdts/stm32mp157-pinctrl.dtsi	2018-11-22 16:28:40.000000000 +0100
+++ ../git/fdts/stm32mp157-pinctrl.dtsi	2018-11-28 10:28:09.336567396 +0100
@@ -134,6 +134,7 @@
 				status = "disabled";
 			};
 
+#if 0
 			uart4_pins_a: uart4@0 {
 				pins1 {
 					pinmux = <STM32_PINMUX('G', 11, AF6)>; /* UART4_TX */
@@ -146,6 +147,21 @@
 					bias-disable;
 				};
 			};
+#else
+			uart4_pins_a: uart4@0 {
+				pins1 {
+					pinmux = <STM32_PINMUX('D', 1, AF8)>; /* UART4_TX */
+					bias-disable;
+					drive-push-pull;
+					slew-rate = <0>;
+				};
+				pins2 {
+					pinmux = <STM32_PINMUX('B', 2, AF8)>; /* UART4_RX */
+					bias-disable;
+				};
+			};
+
+#endif
 
 			usart3_pins_a: usart3@0 {
 				pins1 {
@@ -176,6 +192,7 @@
 				};
 			};
 
+#if 0
 			sdmmc1_dir_pins_a: sdmmc1-dir@0 {
 				pins1 {
 					pinmux = <STM32_PINMUX('F', 2, AF11)>, /* SDMMC1_D0DIR */
@@ -190,6 +207,22 @@
 					bias-pull-up;
 				};
 			};
+#else
+			sdmmc1_dir_pins_a: sdmmc1-dir@0 {
+				pins1 {
+					pinmux = <STM32_PINMUX('F', 2, AF11)>, /* SDMMC1_D0DIR */
+						 <STM32_PINMUX('E', 14, AF8)>, /* SDMMC1_D123DIR */
+						 <STM32_PINMUX('B', 9, AF11)>; /* SDMMC1_CDIR */
+					slew-rate = <3>;
+					drive-push-pull;
+					bias-pull-up;
+				};
+				pins2{
+					pinmux = <STM32_PINMUX('E', 4, AF8)>; /* SDMMC1_CKIN */
+					bias-pull-up;
+				};
+			};
+#endif
 
 			sdmmc2_b4_pins_a: sdmmc2-b4@0 {
 				pins {
diff -Naurw ./fdts/stm32mp15-ddr3-2x4Gb-1066-binG.dtsi ../git/fdts/stm32mp15-ddr3-2x4Gb-1066-binG.dtsi
--- ./fdts/stm32mp15-ddr3-2x4Gb-1066-binG.dtsi	2018-10-18 20:38:12.000000000 +0200
+++ ../git/fdts/stm32mp15-ddr3-2x4Gb-1066-binG.dtsi	2018-11-28 10:23:06.324567396 +0100
@@ -18,8 +18,8 @@
  * address mapping : RBC
  */
 
-#define DDR_MEM_NAME "DDR3-1066 bin G 2x4Gb 533MHz v1.39"
-#define DDR_MEM_SPEED 533
+#define DDR_MEM_NAME "DDR3-1066 bin G 2x4Gb 528MHz v1.00"
+#define DDR_MEM_SPEED 528
 #define DDR_MEM_SIZE 0x40000000
 
 #define DDR_MSTR 0x00040401
@@ -32,20 +32,20 @@
 #define DDR_HWLPCTL 0x00000000
 #define DDR_RFSHCTL0 0x00210000
 #define DDR_RFSHCTL3 0x00000000
-#define DDR_RFSHTMG 0x0081008B
+#define DDR_RFSHTMG 0x0080008A
 #define DDR_CRCPARCTL0 0x00000000
 #define DDR_DRAMTMG0 0x121B2414
-#define DDR_DRAMTMG1 0x000A041C
-#define DDR_DRAMTMG2 0x0608090F
+#define DDR_DRAMTMG1 0x000A041B
+#define DDR_DRAMTMG2 0x0607080F
 #define DDR_DRAMTMG3 0x0050400C
-#define DDR_DRAMTMG4 0x08040608
+#define DDR_DRAMTMG4 0x07040607
 #define DDR_DRAMTMG5 0x06060403
 #define DDR_DRAMTMG6 0x02020002
 #define DDR_DRAMTMG7 0x00000202
 #define DDR_DRAMTMG8 0x00001005
 #define DDR_DRAMTMG14 0x000000A0
 #define DDR_ZQCTL0 0xC2000040
-#define DDR_DFITMG0 0x02060105
+#define DDR_DFITMG0 0x02050105
 #define DDR_DFITMG1 0x00000202
 #define DDR_DFILPCFG0 0x07000000
 #define DDR_DFIUPD0 0xC0400003
@@ -86,37 +86,37 @@
 #define DDR_PCFGWQOS0_1 0x01100B03
 #define DDR_PCFGWQOS1_1 0x01000200
 #define DDR_PGCR 0x01442E02
-#define DDR_PTR0 0x0022AA5B
-#define DDR_PTR1 0x04841104
-#define DDR_PTR2 0x042DA068
+#define DDR_PTR0 0x0022A41B
+#define DDR_PTR1 0x047C0740
+#define DDR_PTR2 0x042D9C80
 #define DDR_ACIOCR 0x10400812
 #define DDR_DXCCR 0x00000C40
 #define DDR_DSGCR 0xF200001F
 #define DDR_DCR 0x0000000B
-#define DDR_DTPR0 0x38D488D0
-#define DDR_DTPR1 0x098B00D8
+#define DDR_DTPR0 0x36D477D0
+#define DDR_DTPR1 0x098A00D8
 #define DDR_DTPR2 0x10023600
-#define DDR_MR0 0x00000840
+#define DDR_MR0 0x00000830
 #define DDR_MR1 0x00000000
 #define DDR_MR2 0x00000208
 #define DDR_MR3 0x00000000
 #define DDR_ODTCR 0x00010000
 #define DDR_ZQ0CR1 0x00000038
 #define DDR_DX0GCR 0x0000CE81
-#define DDR_DX0DLLCR 0x40000000
-#define DDR_DX0DQTR 0xFFFFFFFF
+#define DDR_DX0DLLCR 0x4000C000
+#define DDR_DX0DQTR 0x11101210
 #define DDR_DX0DQSTR 0x3DB02000
 #define DDR_DX1GCR 0x0000CE81
-#define DDR_DX1DLLCR 0x40000000
-#define DDR_DX1DQTR 0xFFFFFFFF
+#define DDR_DX1DLLCR 0x4000C000
+#define DDR_DX1DQTR 0x11101210
 #define DDR_DX1DQSTR 0x3DB02000
 #define DDR_DX2GCR 0x0000CE81
-#define DDR_DX2DLLCR 0x40000000
-#define DDR_DX2DQTR 0xFFFFFFFF
+#define DDR_DX2DLLCR 0x4000C000
+#define DDR_DX2DQTR 0x11101210
 #define DDR_DX2DQSTR 0x3DB02000
 #define DDR_DX3GCR 0x0000CE81
-#define DDR_DX3DLLCR 0x40000000
-#define DDR_DX3DQTR 0xFFFFFFFF
-#define DDR_DX3DQSTR 0x3DB02000
+#define DDR_DX3DLLCR 0x4000C000
+#define DDR_DX3DQTR 0x11101210
+#define DDR_DX3DQSTR 0x3D202000
 
 #include "stm32mp15-ddr.dtsi"
